pkg/kubernetes/reconcile.go

```diff
@@ -14,138 +14,129 @@ import (
+// show/diff/apply cycle. This isn't necessary, but it does help users by producing
+// consistent diffs.
+func Reconcile(raw map[string]interface{}, spec v1alpha1.Spec, kindNameMatchers []*regexp.Regexp) (state manifest.List, err error) {
+	manifests, err := walkJSON(raw)
 	if err != nil {
 		return nil, errors.Wrap(err, "flattening manifests")
 	}
 
-	out := make(manifest.List, 0, len(extracted))
-	for _, m := range extracted {
-		if spec.Namespace != "" && !m.Metadata().HasNamespace() {
-			m.Metadata()["namespace"] = spec.Namespace
+	// If we don't have a namespace, we want to set it to the default that is configured in
+	// our kubernetes specification
+	for _, manifest := range manifests {
+		if spec.Namespace != "" && !manifest.Metadata().HasNamespace() {
+			manifest.Metadata()["namespace"] = spec.Namespace
 		}
-
-		out = append(out, m)
 	}
 
-	// optionally filter the working set of objects
-	if len(targets) > 0 {
-		tmp := funk.Filter(out, func(i interface{}) bool {
-			p := objectspec(i.(manifest.Manifest))
-			for _, t := range targets {
-				if t.MatchString(strings.ToLower(p)) {
+	// If we have any kind-name matchers, we should filter all the manifests by matching
+	// against their <kind>/<name> identifier.
+	if len(kindNameMatchers) > 0 {
+		manifests = funk.Filter(manifests, func(elem interface{}) bool {
+			manifest := elem.(Manifest)
+			kindName := strings.ToLower(manifest.KindName())
+			for _, matcher := range kindNameMatchers {
+				if matcher.MatchString(kindName) {
 					return true
 				}
 			}
+
 			return false
-		}).([]manifest.Manifest)
-		out = manifest.List(tmp)
+		}).([]Manifest)
 	}
 
-	// Stable output order
-	sort.SliceStable(out, func(i int, j int) bool {
-		if out[i].Kind() != out[j].Kind() {
-			return out[i].Kind() < out[j].Kind()
-		}
-		return out[i].Metadata().Name() < out[j].Metadata().Name()
+	sort.SliceStable(manifests, func(i int, j int) bool {
+		return manifests[i].KindName() < manifests[j].KindName()
 	})
 
-	return out, nil
+	return manifests, nil
 }
 
-func extract(deep interface{}) (map[string]manifest.Manifest, error) {
-	extracted := make(map[string]manifest.Manifest)
-	if err := walkJSON(deep, extracted, nil); err != nil {
-		return nil, err
-	}
-	return extracted, nil
+type ErrorPrimitiveReached struct {
+	path      string
+	primitive interface{}
 }
 
-// walkJSON traverses deeply nested kubernetes manifest and extracts them into a flat []dict.
-func walkJSON(deep interface{}, extracted map[string]manifest.Manifest, path trace) error {
-	// array: walkJSON for each
-	if d, ok := deep.([]map[string]interface{}); ok {
-		for i, j := range d {
-			path := append(path, fmt.Sprintf("[%v]", i))
-			if err := walkJSON(j, extracted, path); err != nil {
-				return err
-			}
-		}
-		return nil
-	}
+func (e ErrorPrimitiveReached) Error() string {
+	return fmt.Sprintf("found an object at %s of type %T that was not a valid Kubernetes object", e.path, e.primitive)
```

Comment:
I don't think that this error message is really helpful.

Imagine a new user that just sees:

```
found an object at `.foo.bar` of type `string` that was not a valid Kubernetes object
```

This gives you now clue of what actually happened: While searching for leaf nodes, the end of the tree was reached without finding a valid object.

Reply:
As a user, if I see this, I'd go to foo.bar in my templating and try figuring out why I was producing a non-Kubernetes object. That sounds like what you need to do, right?

Given it's now valid to have non-dict objects in your definition, it wasn't appropriate for the error to say anything about it not being a dict. I think the only thing this misses now is about apiVersion and kind, but that error message was already a bit broken as we should've been checking for name too.

Perhaps this makes more sense if we renamed `s/isKubernetesManifest/validateKubernetesManifest/`, and had it return an error. Then that function, which knows about how it's validating something, can return an appropriate error about what is missing? Otherwise I'd be concerned about writing an error message here about what `isKubernetesManifest` validates that wouldn't necessarily be up-to-date with the validation in the function.

Reply:
The expression `string that was not a valid Kubernetes object` bothers me, as a string will never be a valid object.

I feel like we should tell users at least what we were trying to (find a valid object in the nested output).